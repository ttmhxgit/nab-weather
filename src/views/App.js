import React, { useState, useMemo, useEffect } from "react";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import { debounce } from "../libs/utils";

import NavBar from "../components/NavBar";
import Search from "../components/Search";
import ForeCast from "../components/ForeCast";

const searchTimeoutInMs = 500;

export default function App() {
  const [location, setLocation] = useState("saigon");
  const [debouncedSearchTerm, setDebouncedSearchTerm] = useState("");
  const [isSearching, setIsSearching] = useState(false);

  const debounceSearch = useMemo(
    () =>
      debounce((searchTerm) => {
        setDebouncedSearchTerm(searchTerm);
      }, searchTimeoutInMs),
    []
  );

  const handleLocationChange = (event) => {
    const query = event.target.value.trim();
    if (query) {
      setIsSearching(true);
    }
    debounceSearch(query);
  };

  useEffect(() => {
    if (debouncedSearchTerm) {
      setLocation(debouncedSearchTerm);
      setIsSearching(false);
    }
  }, [debouncedSearchTerm]);


  

  return (
    <div className="min-h-screen">
      <Router>
        <NavBar />
        <Switch>
          <Route exact path="/">
            <main>
              <div className="mx-auto w-full">
                <Search
                  location={location}
                  isSearching={isSearching}
                  onChange={handleLocationChange}
                />
                <div className="shadow-lg rounded-lg h-auto overflow-hidden w-4/5 m-auto mt-4 divide-y-2 divide-light-blue-400">
                  <ForeCast location={location} />
                </div>
              </div>
            </main>
          </Route>
        </Switch>
      </Router>
    </div>
  );
}
