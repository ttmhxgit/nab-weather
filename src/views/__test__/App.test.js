/* eslint-disable jest/no-mocks-import */
import React from "react";
import { cache } from "swr";
import { Router } from "react-router-dom";
import { enableFetchMocks } from "jest-fetch-mock";
import { createMemoryHistory } from "history";
import {
  fireEvent,
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import App from "../App";
import { foreCastWeather } from "../../__mocks__/mock-forecast.mock";

const history = createMemoryHistory();

enableFetchMocks();

const waitForLoadingToFinish = () =>
  waitForElementToBeRemoved(() => [...screen.queryAllByRole(/progressbar/i)], {
    timeout: 4000,
  });

const renderApp = () =>
  render(
    <Router history={history}>
      <App />
    </Router>
  );

describe("app render", () => {
  beforeEach(() => {
    cache.clear();
    fetch.resetMocks();
    fetch.mockResponseOnce(JSON.stringify(foreCastWeather));
  });

  test("fetch and render app", async () => {
    renderApp();

    await waitForLoadingToFinish();

    expect(screen.getByRole("link", { name: /weather/i })).toBeInTheDocument();

    const inputEl = screen.getByPlaceholderText(/type your location/i);
    expect(inputEl).toBeInTheDocument();
    fireEvent.change(inputEl, { target: { value: "New city" } });
    expect(inputEl.value).toBe("New city");

    expect(screen.getByText(/ho chi minh city, vn/i)).toBeInTheDocument();
    const forecast = screen.getAllByRole("listitem").map((listItem) => {
      return listItem.textContent;
    });
    expect(forecast).toMatchSnapshot(`
      Array [
        "MondayH 29°   L 26°",
        "TuesdayH 34°   L 26°",
        "WednesdayH 32°   L 26°",
        "ThursdayH 31°   L 25°",
        "FridayH 31°   L 25°",
        "SaturdayH 31°   L 26°",
      ]
    `);
  });

});
