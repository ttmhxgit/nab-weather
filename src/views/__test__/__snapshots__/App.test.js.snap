// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app render fetch and render app: 
      Array [
        "MondayH 29°   L 26°",
        "TuesdayH 34°   L 26°",
        "WednesdayH 32°   L 26°",
        "ThursdayH 31°   L 25°",
        "FridayH 31°   L 25°",
        "SaturdayH 31°   L 26°",
      ]
     1`] = `
Array [
  "MondayH 29°   L 26°",
  "TuesdayH 34°   L 26°",
  "WednesdayH 32°   L 26°",
  "ThursdayH 31°   L 25°",
  "FridayH 31°   L 25°",
  "SaturdayH 31°   L 26°",
]
`;
